### PR TITLE
Add at::one_hot

### LINKED ISSUE #15060: Add to_one_hot function
## 🚀 Feature
Convert a tensor of index values to one-hot encoding

## Motivation

It's kind of strange not to have it

## Pitch
This old PR does it: https://github.com/pytorch/pytorch/pull/4774

also torch.distributions has it as well, we should just expose it as part of torch.*

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Closes: https://github.com/pytorch/pytorch/issues/15060

cc: @zou3519 

I happen to need this feature, so I just implement it...

This PR contains removal of some unnecessary lambda in `test_torch.py`. These unnecessary lambdas are discovered by automated tools and I just go ahead and fix some of them them. If you don't like this change, I can revert it.

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -1733,6 +1733,9 @@
   variants: method
   device_guard: false
 
+- func: one_hot(IndexTensor self, int64_t num_classes=0) -> Tensor
```
Comment (zou3519) [inline aten/src/ATen/native/native_functions.yaml]: Does `int64_t?` work? It's a little weird to pass in num_classes=0 from the C++ API

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -6241,3 +6241,47 @@ def parse_kwargs(desc):
     >>>                            [7, 8, 9]]))
     (tensor([1, 2, 3]), tensor([4, 5, 6]), tensor([7, 8, 9]))
 """)
+
+add_docstr(torch.one_hot,
+           r"""
+one_hot(tensor, num_classes=0) -> LongTensor
+
+Takes integer with :math:`n` dims and converts it to 1-hot representation
```
Comment (zou3519) [inline torch/_torch_docs.py]: nit: "Takes integer tensor".

Could you add that it takes a tensor of shape `(*)` and returns a tensor of shape `(*, num_classes)` ?

File: torch/distributions/one_hot_categorical.py [hunk 3]
```diff
@@ -76,15 +76,9 @@ def param_shape(self):
...
         indices = self._categorical.sample(sample_shape)
-        if torch._C._get_tracing_state():
-            # [JIT WORKAROUND] lack of support for .scatter_()
-            eye = torch.eye(self.event_shape[-1], dtype=self._param.dtype, device=self._param.device)
-            return eye[indices]
-        one_hot = probs.new_zeros(self._extended_shape(sample_shape))
-        if indices.dim() < one_hot.dim():
-            indices = indices.unsqueeze(-1)
-        return one_hot.scatter_(-1, indices, 1.)
+        return torch.one_hot(indices, num_events).to(probs)
```
Comment (zou3519) [inline torch/distributions/one_hot_categorical.py]: Nice!

File: test/test_torch.py [hunk 4]
```diff
@@ -6686,6 +6686,30 @@ def test_masked_fill(self):
         dst2.masked_fill_(dst2 > 0, val)
         self.assertEqual(dst, dst2, 0)
 
+    def test_one_hot(self):
```
Comment (zou3519) [inline test/test_torch.py]: Can you add a test case for the following two inputs:
1) scalar tensor `torch.one_hot(torch.tensor(1))`
2) tensor with 0-size dim `torch.one_hot(torch.empty([2, 0], dtype=torch.long)`

File: test/test_torch.py [hunk 5]
```diff
@@ -1032,21 +1032,21 @@ def test_dim_reduction(self):
...
+            ('mode', torch.mode, None),
+            ('median', torch.median, None),
+
+            ('prod', torch.prod, 1),
+            ('sum', torch.sum, 0),
+            ('norm', torch.norm, 0),
+            ('mean', torch.mean, nan),
+            ('var', torch.var, nan),
+            ('std', torch.std, nan),
+            ('logsumexp', torch.logsumexp, -inf),
```
Comment (zou3519) [inline test/test_torch.py]: Thanks for the cleanup

**Review by zou3519 — COMMENTED** [review]
Looks great! Minor comments:
- the default behavior of num_classes <= 0 is a little strange, can we use optional<int64_t>? I'm not 100% sure our code generation supports this
- Please add some test cases for some corner cases
- minor documentation nits

File: test/test_torch.py (hunk 4 above)
Comment (zou3519) [reply test/test_torch.py]: (To clarify, these don't have to work, if they don't work we should print out a reasonable error message for them)

**Comment by zasdfgbnm** [comment]
@zou3519 How does this looks like now? The `optional` approach does not to work with JIT, but the feature should be coming soon: https://github.com/pytorch/pytorch/pull/15154

File: aten/src/ATen/native/Onehot.cpp [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+            shape.push_back(num_classes);
+            return at::empty(shape, self.options());
+        }
+    }
+
+    // non-empty tensor
+    if (num_classes <= 0) {
+        num_classes = self.max().item().toLong() + 1;
+    } else {
+        AT_CHECK(self.max().item().toLong() < num_classes, "Class values must be smaller than num_classes.");
```
Comment (zou3519) [inline aten/src/ATen/native/Onehot.cpp]: nit: use AT_ASSERTM to give a better error message:
`AT_CHECK(self.max().item().toLong() < num_classes, "Class values must be smaller than num_classes (", num_classes, "). Got ", self.max().item().toLong(), " class values");

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (zou3519) [inline aten/src/ATen/native/native_functions.yaml]: Let's change the default to `num_classes=-1` and throw on all other non-negative values. `-1` in PyTorch in many cases (view, etc) can mean "infer this number" so I think it is a better value to use than 0

File: test/test_torch.py [hunk 7]
```diff
@@ -6686,6 +6686,45 @@ def test_masked_fill(self):
         dst2.masked_fill_(dst2 > 0, val)
         self.assertEqual(dst, dst2, 0)
 
+    def test_one_hot(self):
+        with self.assertRaises(RuntimeError):
+            torch.one_hot(torch.tensor([3, 4, -1, 0]), -1)
+
+        with self.assertRaises(RuntimeError):
```
Comment (zou3519) [inline test/test_torch.py]: Add test for cuda indices too to `test/test_cuda.py` please

File: aten/src/ATen/native/Onehot.cpp [hunk 8]
```diff
@@ -0,0 +1,35 @@
+#include <ATen/ATen.h>
+
+namespace at { namespace native {
+
+Tensor one_hot(const Tensor &self, int64_t num_classes) {
+    auto shape = self.sizes().vec();
```
Comment (zou3519) [inline aten/src/ATen/native/Onehot.cpp]: Can you check that the type of self is integral and throw an exception if it is not? This makes the error reporting a little nicer

**Review by zou3519 — COMMENTED** [review]
Three more minor things and then this should be good to go!
- Let's change the default to one_hot(self, num_classes=-1), where `-1` means "pytorch, please infer the number of classes from the data"
- check that the input to one_hot is a integral-type tensor 
- add a CUDA test in test/test_cuda.py

**Comment by zasdfgbnm** [comment]
@zou3519 Sorry, I didn't see your suggestion about the `-1` before I finish modifying JIT to support `optional<int64_t>`... A new PR is comming, how do we deal with it?

I will change according to your other suggestions

**Comment by zasdfgbnm** [comment]
@zou3519 How about we keep the -1 as a legal input, and land this PR, then we work on https://github.com/pytorch/pytorch/pull/15235?

I think https://github.com/pytorch/pytorch/pull/15235 would also be useful to finish https://github.com/pytorch/pytorch/pull/12064

**Comment by zasdfgbnm** [comment]
@zou3519 I think landing this one alone (with `num_classes=-1`) is a good idea, because this is much easier in terms of both implementation and review.

File: aten/src/ATen/native/Onehot.cpp (hunk 8 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/Onehot.cpp]: @zou3519 What is the best way to check if it is integral? Currently it looks ugly...

Comment (zou3519) [reply aten/src/ATen/native/Onehot.cpp]: !is_floating_point(self) && !is_complex(self) should work, but it is not the best...

Comment (zou3519) [reply aten/src/ATen/native/Onehot.cpp]: Oh, you could call this macro: https://github.com/pytorch/pytorch/blob/7a61306031648fc4e62fe9e5868893ebf9ae6fb6/aten/src/ATen/Dispatch.h#L54-L66 and have it do nothing; it would check to see if the tensor is integral.

**Comment by vadimkantorov** [comment]
A proposal: does it make sense for `one_hot` to accept a `dim` argument (tensorflow does it with axis) to be created (currently dim is always -1)?

**Comment by zasdfgbnm** [comment]
@zou3519  What is your opinion about @vadimkantorov 's suggestion? Personally I prefer to `torch.one_hot(...).to(...)` rather than `torch.one_hot(..., dtype=)`. I don't know if `dim` would be helpful.

**Comment by zasdfgbnm** [comment]
@zou3519 Done except those suggested by @vadimkantorov

**Review by apaszke — COMMENTED** [review]
Do we really want to add this to the core library? This function is a one-liner in Python...

**Comment by soumith** [comment]
@apaszke it's so common to create one hot encodings, it makes sense to add it. It's not an efficient one-liner in python.

**Comment by fmassa** [comment]
My 2 cents:
- maybe they should live in another namespace, like nn.utils?
- I think it would be nice to have a dtype / device / arguments

**Comment by apaszke** [comment]
How is this less efficient than a native method (and handles all the dtype/device needs as well):
```
def one_hot(classes, num_classes, **kwargs):
    enc = torch.zeros(len(classes), num_classes, **kwargs)
    enc[classes] = 1
    return enc
```

It just seems like a function that doesn't improve the expressive power too much so I'd rather avoid expanding the API unless it gives us measurable gains, but I'm also not super against it, as it is quite common.

**Comment by zasdfgbnm** [comment]
@apaszke I think we need scatter rather than indexing

```python
>>> def one_hot(classes, num_classes, **kwargs):
...     enc = torch.zeros(len(classes), num_classes, **kwargs)
...     enc[classes] = 1
...     return enc
... 
>>> import torch
>>> a = torch.tensor([2,3,4,0,1])
>>> one_hot(a, 6)
tensor([[1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1., 1.]])
```

**Comment by apaszke** [comment]
@zasdfgbnm yeah, I wanted to write scatter of course but have been a bit too quick and haven't even checked if what I wrote works :)

With `scatter_` it's even nicer because it really *is a one liner* to create the zero-filled tensor, and put 1s into those selected places.

**Comment by zou3519** [comment]
@apaszke @soumith I think it's good to add it. A non-zero amount of users will spend some time figuring out how scatter works before understanding how it can be used to implement one-hot, this makes it easier for them to use that functionality

**Comment by zasdfgbnm** [comment]
@zou3519 How does it looks this time? It's still in C++, but now under nn module.

**Comment by vadimkantorov** [comment]
@zou3519 @fmassa `nn.utils` namespace contains only a few advanced functions (weight norm, spectral norm, sequence packing). `one_hot` is for new users, I think if you don't want to clutter the main `torch` namespace, then just `nn` or `nn.functional` is better than `nn.utils`. On the other side, all other factory methods are in `torch`, so maybe it's not that bad to put it in there.

**Comment by soumith** [comment]
no no no. one_hot definitely isn't going in the global `torch` namespace.
Either we do `nn.functional`, or `nn.utils`.

**Comment by zou3519** [comment]
Okay, from the group consensus it sounds like we should toss it into `nn.functional`?

File: aten/src/ATen/native/Onehot.cpp [hunk 9]
```diff
@@ -0,0 +1,35 @@
+#include <ATen/ATen.h>
+
+namespace at { namespace native {
+
+Tensor one_hot(const Tensor &self, int64_t num_classes) {
+    AT_ASSERTM(self.dtype() == kLong, "one_hot is only applicable to index tensor.");
```
Comment (zou3519) [inline aten/src/ATen/native/Onehot.cpp]: AT_ERROR is better, iirc AT_ASSERTM tells the user to submit a bug to pytorch

File: torch/nn/functional.py [hunk 10]
```diff
@@ -2777,6 +2777,55 @@ def pairwise_distance(x1, x2, p=2., eps=1e-6, keepdim=False):
...
+in which case it will be 1.
+
+See also `One-hot on Wikipedia`_ .
+
+.. _One-hot on Wikipedia:
+    https://en.wikipedia.org/wiki/One-hot
+
+Arguments:
+    tensor (LongTensor): class values of any shape.
+    num_classes (int):  Total number of classes. If set to -1, the actually
```
Comment (zou3519) [inline torch/nn/functional.py]: "actually" is not necessary here

File: torch/nn/functional.py [hunk 11]
```diff
@@ -2777,6 +2777,55 @@ def pairwise_distance(x1, x2, p=2., eps=1e-6, keepdim=False):
...
+        number of classes will be inferred as the largest class value in the
```
Comment (zou3519) [inline torch/nn/functional.py]: "the actual number of classes is inferred to be one greater than the largest class value" - the "one greater than" is important

File: aten/src/ATen/native/Onehot.cpp [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+        if (num_classes <= 0) {
+            AT_ERROR("Can not infer total number of classes from empty tensor.");
+        } else {
+            shape.push_back(num_classes);
+            return at::empty(shape, self.options());
+        }
+    }
+
+    // non-empty tensor
+    AT_ASSERTM(self.min().item().toLong() >= 0, "Class values must be non-negative.");
```
Comment (zou3519) [inline aten/src/ATen/native/Onehot.cpp]: Use `AT_ERROR` instead

File: aten/src/ATen/native/Onehot.cpp [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+    if (num_classes == -1) {
+        num_classes = self.max().item().toLong() + 1;
+    } else {
+        AT_ASSERTM(num_classes > self.max().item().toLong(), "Class values must be smaller than num_classes.");
```
Comment (zou3519) [inline aten/src/ATen/native/Onehot.cpp]: Use `AT_ERROR` instead. It might be better for user usability if it were something like
```
AT_ERROR(num_classes > self.max().item().toLong(), "Class values (", self.max().item().toLong(), ") must be smaller than num_classes (", num_classes, ")");
```

**Review by zou3519 — APPROVED** [review]
lgtm, with a few minor nits:

- Let's use AT_ERROR for the error message instead of AT_ASSERTM: AT_ASSERTM is for developers, AT_ERROR is for users
- Please fix the documentation as noted

File: aten/src/ATen/native/Onehot.cpp (hunk 9 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/Onehot.cpp]: @zou3519 Do you mean `AT_CHECK`? I don't think `AT_ERROR` takes any condition as arg, it simply report there is an error unconditionally.

**Review by zou3519 — APPROVED** [review]

**Comment by zou3519** [comment]
@zasdfgbnm test failures look real

**Comment by zasdfgbnm** [comment]
@zou3519 Replacing `AT_ERROR` with `AT_CHECK` should fix the error. Let's see the outcome of CI.

**Comment by zou3519** [comment]
Right, AT_ERROR unconditionally errors but AT_CHECK checks a condition before throwing an error; that is my bad. Thanks for catching that!

**Comment by ezyang** [comment]
Test failures are fake

**Comment by vishwakftw** [comment]
This PR can be closed now, I suppose.

**Comment by zasdfgbnm** [comment]
Why @facebook-github-bot didn't close this?
